controller/runner.go

```diff
@@ -0,0 +1,186 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package controller
+
+import (
+	"fmt"
+	"net"
+	"sync"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+	pb "github.com/elastic/elastic-agent-shipper-client/pkg/proto"
+
+	"github.com/elastic/elastic-agent-shipper/config"
+	"github.com/elastic/elastic-agent-shipper/monitoring"
+	"github.com/elastic/elastic-agent-shipper/output"
+	"github.com/elastic/elastic-agent-shipper/queue"
+	"github.com/elastic/elastic-agent-shipper/server"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+// Output describes a typical output implementation used for running the server.
+type Output interface {
+	Start()
+	Wait()
+}
+
+// ServerRunner starts and gracefully stops the shipper server on demand.
+// The server runner is not re-usable and should be abandoned after `Close` or canceled context.
+type ServerRunner struct {
+	log  *logp.Logger
+	cfg  config.ShipperConfig
+	once sync.Once
```

Comment:
`once` and `mu` should have comments indicating what resources / actions they gate access to. It would also be good to give them names that more explicitly reflect it if there is one, e.g. something called `once` could be gating anything but in this case it looks like it's for shutdown, so `shutdownOnce` or similar might be clearer.